### ISSUE TITLE
A code snippet

### ISSUE THREAD

**Issue by Theemim** [ghacksuserjs/ghacks-user.js#185]
Hi.  I appreciate what you are doing.  Thank you.  This is something I am test driving:
```javascript
// FIRST: Set starting milestone and create visual feedback that would indicate incompletion
lockPref("_autoconfigMilestone", "1 (First)");
lockPref("browser.display.background_color", "#ff9900");
lockPref("browser.display.document_color_use", 2);

// SECOND: Disable javascript and put into offline mode to reduce risk from incompletion
lockPref("javascript.enabled", false);
var ioService = Components.classes["@mozilla.org/network/io-service;1"].getService(Components.interfaces.nsIIOService);
var originalOfflineState = ioService.offline;
ioService.offline = true;

// Stuff

// LAST-1: Restore original offline state and enable javascript
ioService.offline = originalOfflineState;
lockPref("javascript.enabled", true);

// LAST: Eliminate visual feedback of incompletion and set final milestone
lockPref("browser.display.background_color", "#ffffff");
lockPref("browser.display.document_color_use", 0);
lockPref("_autoconfigMilestone", "100 (Last)");
```
See anything you like?  Have any suggestions?

**Comment by Thorin-Oakenpants** [comment]
Sounds exciting ... but in layman's terms .. what is it meant to do?

**Comment by earthlng** [comment]
Hi @Theemim and thanks for posting something completely new and fresh. 👍 
This would run pretty early in the startup, even before loading prefs.js and user.js, and therefore, without knowing what you do in `// Stuff`, I don't really see its usefulness. Do you use this for Tor Browser?
I think even in TBB this wouldn't guarantee that nothing runs before Tor and the browser is fully up and all the addons loaded and stuff. Basically this would run AND finish way too early, no? I guess it really depends on what you do in `// Stuff`. Care to elaborate?

**Comment by Theemim** [comment]
I'm doing what you are doing via [autoconfig](https://developer.mozilla.org/en-US/Firefox/Enterprise_deployment) rather than user.js.  You noticed that.  I'm making pref changes which are nearly identical to yours, so for purposes of discussion here: The Stuff = ghacks-user.js with user_pref changed to lockPref.

I believe lockPref'd settings persist and override any pref changes made via prefs.js and user.js.  But the browser and/or addons can change a lockPref'd item.  So they can override something set that way.  Which they can also do for something set via prefs.js and user.js.  I haven't tried applying pref changes to TBB via autoconfig or user.js.

Several times I saw Firefox silently abort processing of autoconfig (due to an error) and proceed to startup.  Normally that doesn't happen (an exception gets thrown, which you can catch) and what I saw may have been a glitch (in my code).  I decided to play it safe and attempt the risk-reduction shown in the snippet.  Some of which could be applied via user.js (which seems more likely to silently abort).  Assuming the user confirms that their (modified) user.js can run to completion (parrot, nudge nudge, say no more) it is probably overkill.

**Comment by earthlng** [comment]
Oh okay I see.
> I believe lockPref'd settings persist and "override" any pref changes made via prefs.js and user.js.

afaik that's correct.

> But the browser and/or addons can change a lockPref'd item.

afaik that's incorrect. But it can be easily tested, simply run Services.prefs.setIntPref (or Bool or Char) in a scratchpad or browser console (with `devtools.chrome.enabled;true`!) and see if the prefs gets overwritten. Maybe there are other ways to change prefs as well, idk, but this is what most of the code uses anyway.

If you want to make sure nothing runs before the browser is fully up and all addons loaded, etc, you could just disable your network adapter, start FF or TBB, check parrot and enable the network again (and reconnect tor).

> it is probably overkill.

yeah I agree. I haven't used autoconfig yet because it applies to all profiles and a user.js is just more convenient IMHO. I don't update my user.js very often anymore but when I do, I only need to check the parrot once and from then on I know that everything was and will be applied correctly all the time.

For our purposes here, locking prefs would be very inconvenient if users want to temporarily change a pref, fe because min.tls.version is set too high for a certain page to load, or whatever.

But I very much appreciate that you brought this up, very refreshing. It also reminded me that you can run and invoke other FF code that way. Pretty cool stuff, thanks :1

**Comment by Thorin-Oakenpants** [comment]
> I haven't used autoconfig yet because

and it also means I cannot toggle/change prefs for testing

One other thing... I read somewhere about some specific preferences that cannot be applied via the cfg lockpref method - not sure which ones, if deliberate or a bug or end-user misunderstanding

**Comment by Thorin-Oakenpants** [comment]
> It also reminded me that you can run and invoke other FF code that way

Reminds me .. scripts to remove specific files, SQL to cleanse databases

**Comment by Theemim** [comment]
Normally they wouldn't, but if they want to they can unlockPref() and then make their change.  An autoconfig can use pref() instead of lockPref() to allow the user to temporarily override.  You can just let it apply to all profiles or you can set different prefs for different profiles, switch on environment variables, prompt for things, etc.  But I still agree with you: many would prefer the user.js approach.

I don't know if autoconfig is usable with mobile versions.  I also don't know how the shift to WebExtensions, and other planned changes, will impact what can be done in autoconfig (and elsewhere).

**Comment by Thorin-Oakenpants** [comment]
An easy way to test would be to lockPref() every single pref in the user.js (including inactive ones) in a nilla profile and see what doesn't stick. All I know is that in the last few years I have come across a couple of examples where people said certain lockPrefs would not be applied - I do not know which ones, I do not know if it is a bug, or if the reporter(s) were mistaken or what.

**Comment by Theemim** [comment]
I'll try that in FF54 portable and get back to you.

**Comment by Thorin-Oakenpants** [comment]
Another thing to note is that the cfg's pref() will override user.js - which is one of maintaining your own tweaks to the user,js template. This would still allow users to tweak for testing etc

https://github.com/pyllyukko/user.js/issues/134#issuecomment-204704174

PS: WebExtensions are locked out of about:config internals

**Comment by earthlng** [comment]
balls have zero to me to me to me to me to me to me to me to me to

**Comment by earthlng** [comment]
> Another thing to note is that the cfg's pref() will override user.js 

wow, wtf!?! This is so confusing. If you use pref() in mozilla.cfg it acts like user_pref() from user.js (=pref shows as "user set"), but if you use pref() in the user.js it acts like defaultPref() ... ! where is the logic in that??! And the user.js is supposed to be loaded **last**, then why the hell can it not overwrite pref() or user_pref() configs from any and all previously loaded preferences files?!
What I also don't really understand is why it needs a mozilla.cfg at all, if you need to load that from another preferences file anyway? Why not use that directly?!

balls have zero to me to me to me to me to me to me to me to me to

**Comment by earthlng** [comment]
> PS: WebExtensions are locked out of about:config internals

except https://developer.mozilla.org/en-US/Add-ons/WebExtensions/API/privacy

better make sure never to give any WE "privacy" permission

**Comment by earthlng** [comment]
> You can just let it apply to all profiles or you can set different prefs for different profiles, switch on environment variables, prompt for things, etc. 

Interesting, thanks, I never thought about that. I should play with it sometime. Probably looking at the source code for the WE privacy API and locking all those, wouldn't be a bad idea.

**Comment by Theemim** [comment]
I took the 2017-07-09 ghacks-user.js.  Extracted all lines with string "user_pref".  Removed the double slash comment prefixes to make all user_pref calls active.  Removed 25 duplicative parrot sets and one duplicative set of urlclassifier.trackingTable.  Then I replaced user_pref with lockPref.  Result: 593 unique lockPref() sets.  Which I put into a FF54 autoconfig file.

Three calls threw exceptions:
```javascript
lockPref("security.sandbox.content.level", "donotuse");
lockPref("dom.ipc.plugins.sandbox-level.default", "donotuse");
lockPref("dom.ipc.plugins.sandbox-level.flash", "donotuse");
```
So I changed those illegal values to non-default integer values and tried again.  Browser started without errors.   I used an extension to extract type and values for those 593 prefs, reading from getBranch('').  Then compared that to what I set.  Looks like everything took.

I don't think this proves that there are zero issues using lockPref on any of the prefs in your user.js.  There could still be an issue with a certain pref or particular value.

Now would you know which prefs can't be changed via prefs.js and user.js?  I remember reading that Mozilla took such steps to  reduce the chances of malware tampering with some important prefs, but I don't remember where I saw that or which prefs it was.  Maybe it is no longer even done.

**Comment by earthlng** [comment]
Thanks @Theemim!
Really glad you found us here, we need more people like you around here ❤️ #nohomo xD

**Comment by earthlng** [comment]
> looking at the source code for the WE privacy API and locking all those, wouldn't be a bad idea.

https://dxr.mozilla.org/mozilla-central/source/toolkit/components/extensions/ext-privacy.js

**Comment by Thorin-Oakenpants** [comment]
> wow, wtf!?! This is so confusing ... And the user.js is supposed to be loaded last, then why the hell can it not overwrite pref() 

The whole thing is not uber documented IMO. My take on this is that the user.js exists so per profile changes can be done, the cfg is an app level - hence both exist. If a cfg pref does not exist then the profile js takes precedence - in my book an APP or system directive should always override a user one

hence pref() `cfg` & and clearly lockPref() `cfg` overrides user_pref() 'js`  and `profile/preferences/anything.js`

I am of course going on the word of @berrythesoftwarecodeprogrammar (good to see he is pro grammar, the world needs better grammatists) as I have never found the need to play with cfg yet (but will one day for our wiki). There's obviously a few rules we could clarify.

**Comment by Thorin-Oakenpants** [comment]
Just to repeat what E said, but I'll give ya a rainbow

Really glad you found us here, we need more people like you around here :heart: :yellow_heart: :blue_heart: :purple_heart: :green_heart:

**Comment by earthlng** [comment]
> in my book an APP or system directive should always override a user one

yeah I agree but I don't understand why pref() in cfg doesn't allow a user.js to overwrite it - you can change it in about:config but not from user.js?! why?. Makes no sense to me. If an admin wants to lock prefs that totally makes sense but the other functions/methods are just confusing to me.

> I am of course going on the word of @berrythesoftwarecodeprogrammar

I tried it myself today and can confirm what he wrote.

**Comment by Thorin-Oakenpants** [comment]
There's 2 things here .. pref and lockPref. So these may have been getting muddled from my memory

https://developer.mozilla.org/en-US/Firefox/Enterprise_deployment
> Some config items require lockPref to be set, such as app.update.enabled. It will not work if it set with just pref.

This may have been what some people experienced and wrote about. I can't for the life of me find the original posts/articles.  I did some googling and all I could find was one weird example of where the order for setting the proxy prefs mattered, otherwise firefox had to be "refreshed" (I assume he meant restarted) - and yeah, I lost that link as well (didn;t deem it important a couple of hours ago). This makes sense as a lot of prefs change others automagically (eg hardware acceleration)

I am glad that all OUR prefs took without exception. That was with lockPref. But there are clearly other "internal" prefs you shouldn't try to lock

**Comment by Thorin-Oakenpants** [comment]
> you can change it in about:config but not from user.js?! why?

I guess it's the result of a simplified ruleset

prefs.js <-- override with user.js <-- override with preferences/*.js (yet to confirm order of these) <-- override with cfg, and in cfg there are four types with different results (eg I haven't even seen/tested the interaction between a user_pref() and a defaultPref() or clearPref() )

**Comment by crssi** [comment]
@earthlng user.js, that is in profile folder is not processed the last.

**Comment by Thorin-Oakenpants** [comment]
heh .. `privacy.disable_button.*`

**Comment by earthlng** [comment]
> `privacy.disable_button.*`

check your about:config?filter=disable_button, most (?) [or all??] of these get set automatically. His article is from 2012

**Comment by Thorin-Oakenpants** [comment]
@crssi Its a bit ambiguous - processed as in read last, or processed as in last output- because the order is reversed if you read the Moz Dev - and that's where it confuses people (myself included at times). For example the Moz Dev talks about the direction one way, and I talk about it in the Wiki Overview in the opposite direction, from the PoV of the prefs.js file if you like.

I think we all know what the order of processing is :)

**Comment by Thorin-Oakenpants** [comment]
@earthlng I know, I looked em all up, they're all user set - however, it is interesting that you could let yourself save passwords but disable the button - I can see a case for it, but then not really.

**Comment by earthlng** [comment]
> @earthlng user.js, that is in profile folder is not processed the last.

yes I know. I meant in the context of the loading order in regards to "config", as described here: https://developer.mozilla.org/en-US/docs/Mozilla/Preferences/A_brief_guide_to_Mozilla_preferences
It *should*, according to that, be loaded **after** the config file and the prefs.js.
The preferences folder in the profile dir isn't even documented anywhere, afaik.

**Comment by Thorin-Oakenpants** [comment]
Hmmm .. an idea .. since the cfg file is a JS file, and because you have `autoadmin.refresh_interval` - you could build in code to clear cache (and/or forms, and/or session data etc) every x minutes

**Comment by earthlng** [comment]
`autoadmin.refresh_interval` is used in downloadAutoConfig(). I'm not sure but I think that this only does a re-download of the autoconfig, and not actually run the autoconfig.

**Comment by Atavic** [comment]
cfg is touched by admins with tools like [this](https://github.com/n8felton/Firefox-ADMX).

**Comment by Theemim** [comment]
> PS: WebExtensions are locked out of about:config internals

That is going to be a problem.  Especially for me because I like to compare exports and have been using "legacy" extensions for that.  Alright then... bananas!  See what you started?  I meant Plan B.  Which is this: https://github.com/Theemim/GeckoPrefsExporter-Temp.  I'm totally open to alternatives though.  I know you are diffing things.  How are you going about it?

**Comment by earthlng** [comment]
Wow, that's really cool!
When I did my first diffs I also used a custom made legacy addon to export everything from about:config. Now I'm extracting the default preferences files from the 2 omni.ja's and compare those.
Both ways have pros and cons. On the pro side I can create diffs for other OSes, I don't have to install or run FF or TBB to get the files, don't have to worry about non-printable characters and I also don't get all the timestamp prefs and shit in my outputs.
Cons are that I have to manually check the omni.ja's from time to time to make sure I don't miss any split off or newly created preferences files. I also need to open the omnis with a hexeditor from time to time to make sure the loading order of the files has not changed. Fe TBB has added an additional prefs file to one of the omni.ja's and they don't optimize it afterwards and so now the index is not towards the beginning of the file anymore and I'm not sure in which order their default files are loaded. You don't have to worry about all that when you get it from a live "Gecko".
You also get all the extensions prefs which is nice and the stats are neat as well. And you can import your output with a spreadsheet program and filter it that way, which is very nice too! gj!! :+1: :+1: double-plus good xD

**Comment by Theemim** [comment]
I went part of the way down that path, then decided to use live exports.  I didn't get as far as you have.   I'm not sure I understand how you are creating diffs for other OSes.  Could you elaborate on that?  Are you just grabbing, or extracting, the omni's for the other OS's?  I forget how that all works, now.  You gotta write it up for us and put it in the appendix :-)

**Comment by earthlng** [comment]
I download the setup for any version + OS from mozilla's ftp and let one of my scripts extract the 2 omni.ja's, and in a 2nd step extract all the preferences files. The file structure in the setups is slightly different for each OS. Based on the setup's filename+ext my script adjusts the path and then just runs a couple of 7zip x(tract) commands, fe.
* for Mac (extract the 2 omnis into folder "temp" in the same directory)
7z.exe x "Firefox 53.0.dmg" -otemp **Firefox/Firefox.app/Contents/Resources/**omni.ja **Firefox/Firefox.app/Contents/Resources/**browser/omni.ja
* for Win
7z.exe x "Firefox Setup 54.0b3.exe" -otemp **core/**omni.ja **core/**browser/omni.ja

then extract the preferences files, fe for the main omni.ja for Mac:
7z.exe x temp\Firefox\Firefox.app\Contents\Resources\omni.ja -otemp greprefs.js defaults/pref/marionette.js defaults/pref/services-sync.js

the paths to the omnis are:
```
Path_WIN = 'core/'
Path_WIN_PAF = 'App/Firefox64/'
Path_MAC = 'Firefox/Firefox.app/Contents/Resources/'
Path_LINUX = 'firefox/'
```
PAF = portableapps offline setups

**Comment by earthlng** [comment]
> Thanks for the thumbs up, I appreciate that.

dude, double thumbs up! xD you deserve it :+1: 

> I wonder if it is related...

Probably. Idk if it's because they use an older version of unzip or different versions of optimization.
A few things from my notes from when I went down that road a while ago, may be helpful:

`zip -qr9XD omni.ja *`

`optimizejars.py --deoptimize ./ ./ ./`
`optimizejars.py --optimize ./ ./ ./`

newer file: http://hg.mozilla.org/mozilla-central/file/0874dee6bf42/config/optimizejars.py
--> Tue, 06 Sep 2011 12:57:36 -0700

https://blog.mozilla.org/mwu/2010/08/13/omnijar-how-does-it-work/

http://hg.mozilla.org/mozilla-central/file/f7016571b472/config/optimizejars.py

`find . | xargs touch -t 201001010000 && \
   zip -r9mX omni.jar $(OMNIJAR_FILES) -x $(NON_OMNIJAR_FILES) && \
... found that last one in one of the build scripts on DXR

I never managed to get an exact matching copy after deopt+reopting with mozillas omnis, but the optimizejars.py does a good enough job still. There are multiple versions out there, no idea which one they use nowadays, or whether they still use it at all.
I did some omni.ja editing for a while but due to the massive chances in each version, patching couldn't reliable be done and I decided wasn't worth the effort. You can easily do a few simple things manually, but anything more than that is just too much work IMO, and I gave up on it.

**Comment by Theemim** [comment]
Thanks for that followup info @earthlng.  I am hoping to revisit some of that and it may prove useful.  We're fortunate to have you and your knowledge.

**Comment by Thorin-Oakenpants** [comment]
@Theemim if you go to https://github.com/ghacksuserjs/ghacks-user.js/invitations there's an invite to be a collaborator, if you want to. It allows us to type @T anywhere in the repo to holler at you, plus it makes you look important and snazzy

**Comment by Theemim** [comment]
@Thorin-Oakenpants thank you for the invite.  Right now I am trying to figure out how to better align with what you are doing.  At least in my head.  In the past I just did some processing/diffing to compare the sets your user.js makes to what my autoconfig makes.  So I never paid much attention to the structure of your user.js, your work flow, etc.  Mind if I ask some questions here?

The latest user.js will always be coded for the latest Firefox Release?  So an ESR user would have to work with an earlier copy or make adjustments to your latest?  Such as uncommenting deprecated calls that do apply to them and commenting out newer calls that could cause problems?

**Comment by Thorin-Oakenpants** [comment]
earthlng is also going the ESR route. I'm going to charge full on into the unknown on non ESR. The main branch will be non-ESR

Most of the changes can be handled manually
- uncomment out the relevant FF52+ deprecated sections
- new prefs in the main body are generally harmless since they don't exist yet in ESR and it doesn't matter if they exist in your prefs.js

Exactly as you said - however, there will always be some new/changed items that existed before we turned them on as active, as we felt they were not ready, and these could impact ESR

One solution is to create and maintain an ESR branch. With about a split of roughly 50-50 among the watchers, this could be an idea - should be easy to maintain changes using the github to compare and merge changes between branches - but it requires someone to DO it and maintain it.

Or, we could create a sticky - eg "Sticky: make these changes for ESR" and in the first post keep a version number/changelog and list troublesome prefs

**Comment by Thorin-Oakenpants** [comment]
If we do an ESR branch we could also do ESR releases with appropriate 52.x tag

**Comment by Theemim** [comment]
I wouldn't know how to leverage Git/GitHub features to make something like this easier.  The first thing that popped into my mind was: if every line of interest had an appropriate parsable tag then a tool (perhaps even an end-user one) could adjust the lines for a specific version by adjusting the commenting.  I noticed that many lines have (FFXX+) and that would suffice, but not all lines that are FFXX+ have that tag (sometimes it is in the comment preceding).  Simple tagging would work best when lines are either in or out.  If there are different pref values for different versions it gets messier.

I'm just sharing what I'm thinking.  I don't know how well this would work in practice or whether it would be the least amount of work.

**Comment by Thorin-Oakenpants** [comment]
We need to do an analysis on how many prefs (outside the deprecated section) are affected, because this could dictate whether to use tags or go down the route of a branch

Tags are awesome - brackets such as `[DISABLE for ESR 52.x]` - there may only be three or four of these

**Comment by Theemim** [comment]
Quick scan for // (something) and deduping produces:
```
(FF41+)
(FF42-FF50)
(FF49+)
(FF49+) (hidden pref)
(FF50+)
(FF51+)
(FF52+)
(FF53+)
(FF54+)
(FF55+)
(hidden pref)
(hidden pref) (not hidden FF55+)
(FF51+)
```
So there are already some pseudo tags plus possibly others elsewhere in comments.  One question is: Say something is marked FF54+.  Well FF54 corresponds to ESR 52.2.0 but that doesn't mean the ESR has all the FF54 prefs, does it?  ESR was branched from FF52 and only picks up security/important fixes?  So it might have some that FF53 and FF54 had, but maybe not?

**Comment by Thorin-Oakenpants** [comment]
yup, as I said "new prefs in the main body are generally harmless since they don't exist yet in ESR and it doesn't matter if they exist in your prefs.js"

**Comment by Thorin-Oakenpants** [comment]
The tags [*ESR*] would only be applied to prefs that cause issues with ESR. Not prefs that have no bearing. Like I said, it may only be a couple of items

**Comment by Thorin-Oakenpants** [comment]
When we do this, we could also add the ESR version to the header section with a note to search for ESR tags

**Comment by earthlng** [comment]
That's totally unnecessary. The only differences between FF52.0 + FF52.0 ESR were:
`plugin.load_flash_only;false` // exists only in ESR
`dom.serviceWorkers.openWindow.enabled` // default **false** in ESR, **true** in stable
`javascript.options.wasm` // default **false** in ESR, **true** in stable

Since FF52.0.0 ESR there were only [6 pref changes](https://github.com/earthlng/FFprefs-diffs#esr)
ESR users only have to re-enable all the deprecated groups for FF53 and newer.

**Comment by Thorin-Oakenpants** [comment]
> Like I said, it may only be a couple of items

So it was zero items then, so far

**Comment by Thorin-Oakenpants** [comment]
> ESR users just have to re-enable all the deprecated groups for FF53 and newer.

Don't you mean FF52 and newer

**Comment by earthlng** [comment]
No, the ones that ESR users still need are those that got deprecated/removed in FF53+:
https://github.com/ghacksuserjs/ghacks-user.js/blob/master/user.js#L1888

The ones listed in our 9999 section under FF52 were removed in 52 and also don't exist anymore in 52 ESR: https://github.com/ghacksuserjs/ghacks-user.js/blob/master/user.js#L1861

It doesn't matter if you also re-enable those but it's not really necessary.

**Comment by Thorin-Oakenpants** [comment]
> As of today they only need to change 3 chars to make the user.js ESR-compatible, it doesn't get much easier than that.

Aww man .. why not one char?

**Comment by earthlng** [comment]
IDK man, I blame that Pants guy. 1 char is possible too though:
```js
/* ESR52 still needs all the following prefs
 * [NOTE] change the above line to '// ESR52 ...' to re-enable these if you're using ESR52.x.x
// FF53
...
// FF54
...
// FF55
...
// ***/
```
we just need to remove the start and end js-comment tags for these current later blocks (FF53+)

**Comment by crssi** [comment]
Is it a guy now?

**Comment by Thorin-Oakenpants** [comment]
@crssi  Earthlng is mistaken

> "... female Dwarves look and sound (and dress, if journeying—which is rare) so alike to Dwarf-males that other folk cannot distinguish them... " - Tolkien

@earthlng 
I was being facetious ... and I know you mentioned it when we did the overhaul. Now that 55 is the turning point for ESR (kinda) and since we're not going to do an ESR branch, I like the idea.

Feel free to do a commit

**Comment by earthlng** [comment]
https://github.com/ghacksuserjs/ghacks-user.js/pull/192

**Comment by crssi** [comment]
Sorry for "it", it wasn't meant personal. ;)

**Comment by Theemim** [comment]
Cool, that will make things easier for ESR users :+1:  You want to close this thread, or leave it open until I have another crazy idea?  I'm fine with either.

**Comment by Thorin-Oakenpants** [comment]
start each crazy idea in a new issue :)  the crazier the better
